##Code below is for processing fastq files from the Superinfection Reservoir Study using R markdown

```{r, include = FALSE}
library(dada2)
library(Rcpp)
library(data.table)
library(seqinr)
library(ape)
library(ggtree)
library(phangorn)
library(stringr)
library(msa)
library(data.table)
library(tools)
library(cowplot)
library(gridExtra)
library(ggplot2)
library(ggtree)
library(treeio)
library(plyr)
```

## This chunk cycles through each case and processes the raw reads. Change the primer.fwd variable in filterAndTrim to the region you are interested in (Gag or Env).
```{r, include=FALSE}
Sample <- c("QB045",
            "QB726",
            "QC885",
            "QD022",
            "QF564",
            "QG262")

##Place fastq files for each case in their own folder in a directory and name the folders according to the sample.  Change the code to reflect the location you have placed the folders.  
for(x in Sample){
setwd(paste("~/NGS_Analysis/QRS_Reservoir_Paper_Illumina_Only_2019/",x, sep =""))
path <- paste("~/NGS_Analysis/QRS_Reservoir_Paper_Illumina_Only_2019/", x, sep = "") 
list.files(path)

# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq

fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)

filt_path <- file.path(path, "#") #Place filtered files in filtered/ subdirectory
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sample.names, "_R_filt.fastq.gz"))
#The below filters v3 sequences
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs,
                     rm.phix=TRUE,
                     compress=TRUE,
                     multithread=4,
                     primer.fwd =  "TACATTATTGTGC",  #primer fwd to pull out only sequences of region
                     matchIDs = TRUE,
                     maxEE = 0.5,
                     minLen = 250)

# Gag fwd primer "CAGGGAGCTGGAAAGATTYGC"
# v3 fwd primer "TACATTATTGTGCYCCAGCTGG"


write.csv(out, file = "#.csv")

errF <- learnErrors(filtFs, multithread=4)
errR <- learnErrors(filtRs, multithread=4)
plotErrors(errF, nominalQ=TRUE)

derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs) <- sample.names
names(derepRs) <- sample.names


dadaFs <- dada(derepFs, err=errF, multithread=4, selfConsist = TRUE, pool = TRUE, OMEGA_C = 0,
               OMEGA_A = 1e-10, USE_QUALS = TRUE)
dadaRs <- dada(derepRs, err=errR, multithread=4, selfConsist = TRUE, pool = TRUE, OMEGA_C = 0,
               OMEGA_A = 1e-10, USE_QUALS = TRUE)

dadaFs[[1]]
dadaRs[[1]]


mergers <- mergePairs(dadaFs,
                      derepFs,
                      dadaRs,
                      derepRs,
                      verbose=TRUE,
                      justConcatenate = FALSE,
                      maxMismatch = 5) 

seqtab <- makeSequenceTable(mergers)
dim(seqtab)
table(nchar(getSequences(seqtab)))
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(mergers, getN), rowSums(seqtab), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoised", "merged", "tabled", "nonchim")
track
final_df <- as.data.frame(t(seqtab))

View(final_df)
write.csv(final_df, file = paste(x,"#.csv"))
sequences = setDT(final_df, keep.rownames = TRUE)[]
sequences <- as.data.frame(sequences)
sequences <-  sequences[, colSums(sequences != 0) > 0]
sequences = setDT(sequences, keep.rownames = FALSE)
sample.names  <- as.character(colnames(sequences[,2:ncol(sequences)]))

final <- NULL
#Only want sequences that are greater than 0.5% in abundance
for(sample in sample.names) {
  c <-  assign(paste("Env", sample, sep =""), data.frame(cbind(sequences$rn[sequences[, get(sample)] > sum(sequences[, get(sample)])*0.005],
  sequences[, get(sample)][sequences[, get(sample)] > sum(sequences[, get(sample)])*0.005], sample)))
  final <- rbind(final,c)
}


samplename <- read.csv("SampleList.csv")

names(final) <- c("sequence", "count", "Sample_ID")
final <- merge(final, samplename, by = "Sample_ID")
final$unique <- seq.int(1,nrow(final),1)

final$Sequence.Name <- paste(final$unique, final$Sequence.Name, final$count, sep = "_")


final <- data.frame(final$sequence,
                    final$count,
                    final$Sequence.Name)
names(final) <- c("sequence", "count", "name")
write.csv(final, file = paste(x,"nochim_Sequence_Count_Name_Env.csv", sep = "_"))


ref <- readDNAStringSet("ref.v3.fasta", format = 'fasta')
ref
ref1 <- as.data.frame(ref)
names(ref)
ref1$count = as.numeric(1)
ref1$name <- names(ref)
names(ref1) <- c("sequence", "count", "name")


Env <- data.frame(final)
Env$name <- as.character(Env$name)
Env$count <- as.numeric(as.character(Env$count))
n_occur <- data.frame(table(Env$sequence))
n_occur[n_occur$Freq > 1,]


##I want to dereplicate sequences that are duplicates at that specific timepoint.

c <- as.data.frame(str_split_fixed(Env$name, "_", 6))

Env$timepoint <- c$V4
Env$Nucleic_Acid <- c$V5

dereplicated <- as.data.frame(ddply(Env,~sequence + timepoint + Nucleic_Acid ,summarise,count=sum(count)))
dereplicated$unique <- seq.int(1,nrow(dereplicated),1)

dereplicated$name <- paste(dereplicated$unique,
                           x,
                           "Env",
                           dereplicated$timepoint,
                           dereplicated$Nucleic_Acid,
                           dereplicated$count,
                           sep = "_")

f = as.data.frame(cbind(as.character(dereplicated$sequence),
                   dereplicated$count,
                   dereplicated$name))
names(f) <- c("sequence",
              "count",
              "name")

Env <- data.frame(rbind(f,
                             ref1))

Env$name <- as.character(Env$name)


write.fasta(sequences = as.list(Env$sequence),
            names = as.list(Env$name), paste(x, "#.fasta", sep = ""))

}

```

